src/hotspot/share/runtime/handshake.cpp

```diff
@@ -327,7 +327,7 @@ void HandshakeOperation::do_handshake(JavaThread* thread) {
   // here to make sure memory operations executed in the handshake
   // closure are visible to the VMThread/Handshaker after it reads
   // that the operation has completed.
-  Atomic::dec(&_pending_threads, memory_order_release);
+  Atomic::dec(&_pending_threads);
```

Comment:
Why do we need a full fence instead of a release as before?

Reply:
We have three options here:
1: "If the handshake operation is not completed it must be in at least one queue."
2: "If the handshake operation is completed it is not on any queue."
3: "It may be on queue while completed and may not be queue when not completed, it depends."

With the polling mechanism saying you must stop while the queue is not empty, the operation is removed after it is executed, option 1 seemed preferable to me.
Specially if we want a simpler easier model to avoid any further such bugs as this.

This means removal from queue happens after decrement of _pending_threads.
remove_op() -> pop() gives no ordering guarantee, those comes from the mutex.

The technical note on the current implementation of the queue is:
Only first can be read without lock and the implementation of first manipulation is done with a CAS.
So a side-effect of current implementation is that the only thing that can be seen will be ordered.

Since we running in slowpath there is no need to reduce ordering to a minimum and we can thus change the queue implementation since it have no guarantees to do any ordering without worrying about someone (ab)using side-effect.

Do you agree?

Reply:
Right, but I don't see why the change of just using the queue to check for pending handshakes matters here. This release was to ensure that if the handshakee executed the closure, the memory operations executed in the handshake would be visible to the vmthread/handshaker after seeing the operation was completed (matched with the acquired after is_completed()). What could be the issue if this is not a full fence?

Reply:
"Change of just using the queue to check for pending handshakes": also means we must remove the handshake from the queue after it have been executed.
I consider the decrement as the marking of "it have been executed".

Therefore anything after the decrement, should stay after and not float above the dec, to follow the protocol.
If disregard ordering side-effects when removing the last in queue with CAS, if the removal happens before the decrement means if the handshakee ask "should process" it could say no before we decrement.

This release to the removal of the operation is match with the acquire on is_empty().
Normally the removal would had used the release from the mutex lock, but here we hold the lock over several operations.

Reply:
Ok, although since all it matters for is_empty() is that there is at least one element in the queue and to remove the last element we always use a CAS it should be okay. I now realize that we could have also removed the operation from the queue as before if it is not the last one, so that would save the second traversal of the queue. But I see that if we want to maintain the invariant "If the handshake operation is not completed it must be on at least one queue" then we need to do it like this.

Reply:
Did you guys decide the comment needed updating or not?
I can't figure that out from the above.

Reply:
Maybe it should be added in the comment that the fence is also used to prevent removing of the operation to float above marking the operation as completed so we keep the invariant: "If the handshake operation is not completed it must be on at least one queue".

Reply:
I still think we can keep the previous memory_order_release order, but if this stricter order makes it easier to read for others then I'm okay.

Reply:
As long as we have a release() fence after the handshake closure the reordering of decrementing _pending_threads and removing the operation from the queue shouldn't matter since the target doesn't read _pending_threads.

Reply:
By release() fence I was referring to the definition in orderAccess (LoadStore | StoreStore), sorry for the confusion, which would be enough to prevent any of those writes to float above the barrier. I now see that dec( ,memory_order_release) might not necessarily provide that though so we would need to use OrderAccess::release() + dec( ,memory_order_relaxed) instead. But I see that using a full fence is probably easier to read. Also by looking at the implementations of dec() for arm32/aarch64 it seems the order parameter is ignored (always full fence) (?), which would defeat the purpose of using release().